File: Design-Patterns/mvp/app/src/main/java/com/example/mvc/data/network/Network.java
package com.example.mvc.data.network;

import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Network {

    private ProductService productService;
    private static Network instance;
    private static Retrofit retrofit;

    private Network(){}
    public static Network getInstance(){
        if (instance == null){
            instance = new Network();
            retrofit = new Retrofit.Builder()
                    .baseUrl("https://dummyjson.com/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                    .build();
        }
        return instance;
    }

    public ProductService getProducts(){

        productService = retrofit.create(ProductService.class);
        return productService;
    }
}

File: Design-Patterns/MVVM/app/src/main/java/com/example/mvvm/data/network/Network.kt
package com.example.mvvm.data.network

import retrofit2.Retrofit
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

object Network {
    private const val BASE_URL = "https://dummyjson.com/"

    val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }
}